### ISSUE TITLE
方案评审：5 折按人交叉验证（exp8）——W+P 对照 + 空间可靠性诊断

### ISSUE THREAD

**Issue by zjucgl** [zjucgl/triscalp#2]
> 状态：**方案待评审**，未实现。评审通过后开发 `experiments/exp8_crossval.py`。

## 实验目标（三个问题）

- **Q1（主问题）**：第三个光源（UV）值不值得？对照是 **W+P 双光源方案**，不是最佳单模态——论文要回答的是"第三个光源的边际价值"。
- **Q2（嵌套消融）**：空间可靠性项在全局项之上有无真实增量？（triscalp_g vs triscalp_gs）
- **Q3（机制诊断）**：空间可靠性估计器 R_s 是否真的响应高光区？现有证据不足：高光区内 A_W 0.405 vs 区外 0.419，差仅 0.014；而物理上高光区毛发对比度近乎归零，真正工作的空间项应把该区 A_W 压到接近 0。

## 一、折构造

- 数据：`aligned_gt/`（60 样本，manifest `subject` 列 32 人），全分辨率 1280×720
- **按 subject 分 5 折**（同人所有部位同折防泄漏，按样本数均衡，固定种子）
- 轮转：第 k 折 test（≈12），第 (k+1)%5 折 val（≈12，best-checkpoint 选择），其余 train（≈36）
- 每折模态标准化统计量**仅用该折 train** 计算（float64）

## 二、方法集（8 × 5 折 = 40 次训练）

| # | 方法 | 模态 | 作用 |
|---|---|---|---|
| 1 | Concat-WP | W+P | 双光源对照（最强朴素基线） |
| 2 | TriScalp-WP | W+P | 同机制双光源版，分离"加 UV"净效应 |
| 3 | Average-WPU | W+P+U | 等权负对照（预期被 UV 拖累） |
| 4 | Concat-WPU | W+P+U | 可学习三光源 |
| 5 | Attention-WPU | W+P+U | 全局自适应 |
| 6 | TriScalp-G | W+P+U | 仅 Global（嵌套基） |
| 7 | TriScalp-GS | W+P+U | Global+Spatial（嵌套增量） |
| 8 | TriScalp-Full | W+P+U | G+S+Task 完整版 |

训练配置沿用论文口径：60 epochs、batch 8、cosine、best-val、人工模态退化增强 p=0.5、固定 seed。

## 三、评估条件（每折 test）

- **Clean**：Dice / IoU / clDice / BF1
- **Degraded-WP**：W@L2 与 P@L2 均值（全部方法可公平比较）
- **U-L2** 单列（仅对 WPU 方法有意义，不混入平均）

## 四、关键对比与统计（逐折配对）

| 对比 | 回答 |
|---|---|
| Δ1 = TriScalp-Full − Concat-WP | 第三光源 + 机制的总收益（论文主结论） |
| Δ2 = TriScalp-Full − TriScalp-WP | **同机制下加 UV 的净收益（最干净的 Q1 答案）** |
| Δ3 = TriScalp-GS − TriScalp-G | 空间项净增量（Clean / Degraded 各报） |
| Δ4 = Concat-WPU − Concat-WP | 朴素机制下加 UV 收益（与 Δ2 对照） |

报告 5 折 mean±std + 逐折数值 + paired t-test 与 Wilcoxon（n=5 功效有限，如实标注）。

## 五、空间项机制诊断（Q3）

对 TriScalp-GS / Full 在各折 test 样本上：

1. 高光代理掩码（白光亮度 > 0.92）内 vs 外统计 **R_s^W 原始 sigmoid 输出**——比 softmax 后的 A 更直接反映估计器是否学到"高光=不可靠"（A 被 r_g、T、softmax 稀释）；
2. 同时报告区内/外 A_W 与 R_s 对照；
3. 解读预案：
   - R_s 区内外无差 → 空间项确实没学到，下一步做 **patch 级局部退化 + R_s 显式监督**（把 §16 的全局 q 监督推广到空间）；
   - R_s 有差但 A 无差 → softmax/量纲稀释问题，改进方向是可学习 β 或解耦 r_g 与 R_s 的量纲。

代码改动：`TriScalpFusion.info` 暴露 R_s 原始图（一行级，不影响既有实验）。

## 六、产出与开销

- `results_cv/table_cv_main.md`（mean±std）、`table_cv_deltas.md`（Δ1–Δ4 + p 值）、`spatial_diagnosis.json`、逐折明细、汇总报告
- 40 次训练 × 全分辨率 60 epochs ≈ A100 **2–3 小时** + 评估诊断约 20 分钟

## 待确认

- [ ] 方法集 8 个是否合适？（可砍 Average-WPU / Attention-WPU 省 10 次训练，代价是少负对照）
- [ ] 高光代理阈值：白光亮度 > 0.92 是否合理？有无真实镜面反射标注可用？
- [ ] 模型选择：留 val 折做 best-checkpoint（train=36），还是不留 val 跑满 60 epochs 取末代（train=48）？

**Comment by zjucgl** [comment]
## 实现完成，可运行

按方案实现并通过本地端到端冒烟（迷你数据 2 折 × 2 epochs 全流程）+ 39 项单元测试。已部署到服务器。

**三个待确认点按默认执行**（后续可改）：完整 8 方法集；高光代理 = 白光亮度 > 0.92（`exp8_crossval.py` 顶部 `GLARE_THRESHOLD` 可调）；保留 val 折做 best-checkpoint（train=36/折）。

### 运行指令（服务器 `/usr/cgl/triscalp`）

```bash
cd /usr/cgl/triscalp

# ① CV 数据转换（全分辨率，一次性，约 5 分钟）
python3 -m triscalp.data.aligned --src aligned_gt --out data_cv --scale 1 --cv-folds 5

# ② 5 折 CV（8 方法 × 5 折 = 40 次训练，A100 约 2-3 小时；支持中断续跑）
nohup python3 -u -m experiments.exp8_crossval --data-root data_cv --out results_cv --epochs 60 \
  > /root/cv.log 2>&1 &

tail -f /root/cv.log          # 看进度
```

### 产出（`results_cv/`）

- `table_cv_main.md`：8 方法 × Clean / Deg-WP / U-L2（mean±std）
- `table_cv_deltas.md`：Δ1–Δ4 配对对比 + 逐折数值 + paired-t / Wilcoxon p 值
- `summary.json`：含 `spatial_diagnosis`（R_s^W 与 A_W 的高光区内/外对比，Q3 答案）
- 日志末尾直接打印 Δ 表与空间诊断摘要

### 实现要点

- 折按 subject 均衡贪心分配（固定种子，样本数 4/4/4/3/3 级均衡），同人不跨折
- 每折标准化统计量只用该折 train（float64 流式计算，无精度饱和）
- 逐 (fold, method) 缓存 checkpoint 与指标，中断后重跑自动跳过已完成项
- `TriScalpFusion` 新增暴露 R_s 原始 sigmoid 图（诊断用，不影响既有实验）

**Comment by zjucgl** [comment]
## ⚠️ 缺陷记录：当前 CV 存在 subject 泄漏

对照 `formal_subject_folds_v2.csv` / manifest v2 发现：`aligned_gt/manifest.csv` 的 `subject` 列实为 **session_id**（32 个），真实 `subject_id` 仅 18 个。exp8 按该列分折导致 **5/18 真实 subject 跨折**（最严重者跨 4 折）。

**影响评定**：
- ❌ Δ1–Δ4 的绝对数值与 p 值：偏乐观，不可用作论文口径
- ✅ 空间诊断（Q3，R_s 高光响应）：机制属性，与折划分无关，仍有效
- ✅ 方法间相对比较：同等受影响，方向性可作预览
- ✅ 管线工程验证有效

**处理**：当前运行跑完取 Q3 与预览（结果标注泄漏警告）；分折逻辑修复（改用 manifest v2 `subject_id` / 官方 folds v2）并入 #3 Track F 实现；协议干净的 Δ 结论以 Track F（官方折）为准。

**Comment by zjucgl** [comment]
## CV 完成：Q3 诊断结论 + Δ 预览（泄漏警告）

### Q3 空间诊断（有效，不受泄漏影响；n=60 样本）

| | R_s^W 高光内 | 高光外 | Δ | A_W 高光内 | 高光外 |
|---|---|---|---|---|---|
| TriScalp-GS | 0.386 | 0.448 | −0.062 | 0.318 | 0.331 |
| TriScalp-Full | 0.409 | 0.470 | −0.062 | 0.397 | 0.413 |

**结论：R_s 方向正确但响应太弱（−0.06，远非压至 0），且经 softmax 稀释后 A 仅差 0.014。**
改进路径：① patch 级局部退化 + R_s 显式监督；② 可学习 β/量纲解耦。是否实施待 Track F（#3）官方折结果确认 D3 后决定。

### Δ 预览（⚠️ subject 泄漏 + 非官方折，仅方向参考，正式版见 #3 Track F）

D1 Full−ConcatWP = −0.003 (ns)；**D2 Full−TriScalpWP = +0.0002 (ns，同机制加 UV 对分割零增益)**；
D3 GS−G = +0.002 (ns)；D4 ConcatWPU−ConcatWP = +0.005 (p≈0.05，边缘)。主表最优 Concat-WPU 0.8865±0.013。

**方向性含义**：UV 对毛发分割边际价值趋近于零 → 三光源价值主张应落在状态任务（Track S）；
"可靠性融合保证弱模态不拖累主任务 + 为状态任务保留 UV" 或成为更诚实的论文叙事。
